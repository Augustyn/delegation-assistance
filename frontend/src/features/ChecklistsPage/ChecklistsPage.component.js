import React from "react";
import { array, bool, func } from "prop-types";
import { reduxForm, Form } from "redux-form";

import LayoutMain from "../../components/layouts/LayoutMain/LayoutMain.container";
import Button from "../../components/Button/Button.component";
import Card from "../../components/Card/Card.component";
import CheckboxGroupChecklist from "../../components/CheckboxGroupChecklist/CheckboxGroupChecklist.component";

export const ChecklistsPage = props => {
  const { tasks = [], handleSubmit, handleDelete, isUserTravelmanager } = props;
  return (
    <LayoutMain title="Checklist">
      <Card>
        <Form onSubmit={handleSubmit}>
          <div className="checklist-list">
            <CheckboxGroupChecklist options={tasks} path="id" name="tasks" />
          </div>
          <Button type="submit" text="Save" />
          {isUserTravelmanager ? (
            <span className="delete-wrapper">
              <Button
                type="submit"
                text="delete"
                className="warning"
                onClick={handleSubmit(values => handleDelete(values))}
              />
            </span>
          ) : null}
        </Form>
      </Card>
    </LayoutMain>
  );
};

ChecklistsPage.propTypes = {
  tasks: array,
  isUserTravelmanager: bool,
  handleSubmit: func,
  handleDelete: func
};

export default reduxForm({
  form: "updateTasks"
})(ChecklistsPage);
