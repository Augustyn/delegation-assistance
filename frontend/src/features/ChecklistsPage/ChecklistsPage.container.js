import React, { Component } from "react";
import { connect } from "react-redux";
import { array, bool, func } from "prop-types";

import ChecklistsPage from "./ChecklistsPage.component";
import { getTasks } from "../../selectors/checklists.selectors";
import { deleteTask, fetchTasks, saveTasks } from "../../actions/checklists.actions";
import { isUserTravelmanager } from "../../selectors/user.selectors";

class ChecklistsPageContainer extends Component {
  static propTypes = {
    tasks: array,
    fetchTasks: func,
    saveTasks: func,
    deleteTask: func,
    isUserTravelmanager: bool
  };

  componentDidMount = () => {
    this.props.fetchTasks();
  };

  _handleSubmit = values => {
    this.props.saveTasks(values);
  };

  _handleDelete = values => {
    if (values) {
      values.tasks.forEach(task => {
        this.props.deleteTask(task);
      });
    }
  };

  render() {
    return (
      <ChecklistsPage
        tasks={this.props.tasks}
        onSubmit={this._handleSubmit}
        handleDelete={this._handleDelete}
        isUserTravelmanager={this.props.isUserTravelmanager}
      />
    );
  }
}

const mapStateToProps = state => {
  return {
    tasks: getTasks(state),
    isUserTravelmanager: isUserTravelmanager(state)
  };
};

const mapDispatchToProps = { deleteTask, fetchTasks, saveTasks };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ChecklistsPageContainer);
