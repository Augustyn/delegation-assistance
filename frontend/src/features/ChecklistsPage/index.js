import ChecklistsPage from "./ChecklistsPage.container";
export default ChecklistsPage;
