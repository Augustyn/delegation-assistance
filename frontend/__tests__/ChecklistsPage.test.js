import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { ChecklistsPage } from "../src/features/ChecklistsPage/ChecklistsPage.component";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

Enzyme.configure({ adapter: new Adapter() });

const setup = () => {
  const props = {
    handleSubmit: jest.fn(),
    subbmiting: false,
    invalid: false,
    tasks: [
      {
        id: 1,
        name: "Checkbox 1",
        description: "Description 1"
      },
      {
        id: 2,
        name: "Checkbox 2",
        description: "Description 2"
      },
      {
        id: 3,
        name: "Checkbox 3",
        description: "Description 3"
      }
    ],
    isUserTravelmanager: true,
    handleDelete: jest.fn()
  };

  const enzymeWrapper = shallow(<ChecklistsPage {...props} />);

  return {
    props,
    enzymeWrapper
  };
};

describe("Checklists Page", () => {
  let enzymeWrapper;
  let props;

  beforeEach(() => {
    ({ enzymeWrapper, props } = setup());
  });

  it("should render correctly", () => {
    expect(enzymeWrapper).toMatchSnapshot();
  });

  it("When user WITHOUT role manage is logged in, should render second button", () => {
    const state = { ...props, isUserTravelmanager: false };
    const WithoutManagerRole = shallow(<ChecklistsPage {...state} />);
    expect(WithoutManagerRole.find("Button").length).toBe(1);
  });

  it("When user WITH role manage is logged in, shouldn't render second button", () => {
    expect(enzymeWrapper.find("Button").length).toBe(2);
  });
});
