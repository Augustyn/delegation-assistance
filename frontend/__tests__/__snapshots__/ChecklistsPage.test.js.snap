// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checklists Page should render correctly 1`] = `
<Connect(LayoutMainContainer)
  title="Checklist"
>
  <Card>
    <Form
      onSubmit={
        [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
    >
      <div
        className="checklist-list"
      >
        <CheckboxGroupChecklist
          name="tasks"
          options={
            Array [
              Object {
                "description": "Description 1",
                "id": 1,
                "name": "Checkbox 1",
              },
              Object {
                "description": "Description 2",
                "id": 2,
                "name": "Checkbox 2",
              },
              Object {
                "description": "Description 3",
                "id": 3,
                "name": "Checkbox 3",
              },
            ]
          }
          path="id"
        />
      </div>
      <Button
        text="Save"
        type="submit"
      />
      <span
        className="delete-wrapper"
      >
        <Button
          className="warning"
          text="delete"
          type="submit"
        />
      </span>
    </Form>
  </Card>
</Connect(LayoutMainContainer)>
`;
